SQL: Implement to_char function (#54964) (#63766)
elastic/elasticsearch: docs/reference/sql/functions/date-time.asciidoc

include-tagged::{sql-specs}/docs/docs.csv-spec[formatTime]
--------------------------------------------------

[[sql-functions-datetime-to_char]]
==== `TO_CHAR`

.Synopsis:
[source, sql]
--------------------------------------------------
TO_CHAR(
    date_exp/datetime_exp/time_exp, <1>
    string_exp) <2>
--------------------------------------------------

*Input*:

<1> date/datetime/time expression
<2> format pattern

*Output*: string

*Description*: Returns the date/datetime/time as a string using the format specified in the 2nd argument. The formatting

> pattern used is one from

Reviewer: ```suggestion
patterns conform to
```
since multiple pattens can be used, not just one.

> https://www.postgresql.org/docs/9.1/functions-formatting.html#FUNCTIONS-FORMATTING-DATETIME-TABLE[PostgreSQL Template Patterns for Date/Time Formatting]

Reviewer 2: Hm, not really, probably that was the current version of postgres back then, when the issue was opened.

> or, if the pattern is not found, from Java

Reviewer: ```suggestion
and to Java
```
plus a trailing `in that order, for overlapping patterns in the two specifications`. Or something more along these lines.

https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/time/format/DateTimeFormatter.html[Pattern Letters and Symbols].
If any of the two arguments is `null` or the pattern is an empty string `null` is returned.

[NOTE]
If the 1st argument is of type `time`, then pattern specified by the 2nd argument cannot contain date related units
(e.g. 'dd', 'MM', 'YYYY', etc.). If it contains such units an error is returned.

*Special Cases*

- Postgres formats `IDDD` and `WW` are not covered.

> - Format Specifier `y` will return year-of-era instead of one/two low-order digits.

Reviewer: ```suggestion
- Format Specifier `y` will return year-of-era (conforming to the Java specification) instead of the last digit of the year.
```
(the `two low-order digits` will work as expected, same for both specs, right?)

I believe we should also mention that `Y` will work as per the Postgresql specs, especially since all Postgresql patterns are case insensitive, except where made explicit othewise.
